geppetto.js/geppetto-ui/src/3d-canvas/threeDEngine/ThreeDEngine.js

```diff
@@ -198,17 +198,141 @@ export default class ThreeDEngine {
+      for (let child of mesh.children) {
+        if (this.checkMaterial(child, instance)) {
+          changed = true;
+        }
+      }
+      return changed;
+    }
+  }
+
+  updateInstanceMaterial(mesh, instance) {
+    if (mesh.type === 'Mesh') {
+      if (instance?.color?.r !== undefined
+        && instance?.color?.g !== undefined
+        && instance?.color?.b !== undefined
+        && instance?.color?.a !== undefined) {
+        for (let child of this.scene.children) {
+          if (child.instancePath === mesh.instancePath && child.uuid === mesh.uuid) {
+            child.material.color.setRGB(instance.color.r, instance.color.b, instance.color.b);
+            if (instance.color.a) {
+              child.material.transparent = true;
+              child.material.opacity = instance.color.a;
+            }
+          }
+        }
+      } else {
+        console.error("color for instance provided is malformed, please provide the color in the form");
+        console.error("{r: [0-1], g: [0-1], b: [0-1], a: [0-1]}");
+      }
+    } else if (mesh.type === 'Group') {
+      if (instance?.color?.r !== undefined
+        && instance?.color?.g !== undefined
+        && instance?.color?.b !== undefined
+        && instance?.color?.a !== undefined) {
+        for (let child of this.scene.children) {
+          if (child.instancePath === mesh.instancePath && child.uuid === mesh.uuid) {
+            // TODO: can I have nested Group instances? to check and fix in case
+            for (let granChild of child.children) {
+              granChild.material.color.setRGB(instance.color.r, instance.color.b, instance.color.b);
+              if (instance.color.a) {
+                granChild.material.transparent = true;
+                granChild.material.opacity = instance.color.a;
+              }
+            }
+          }
+        }
+      } else {
+        console.error("color for instance provided is malformed, please provide the color in the form");
+        console.error("{r: [0-1], g: [0-1], b: [0-1], a: [0-1]}");
+      }
+    }
+  }
+
   /**
    * Clears the scene
    *
    */
-  clearScene () {
-    const toRemove = this.scene.children.filter(
-      child => child.type === 'Mesh'
+  clearScene (instances) {
+    let topInstances = [];
+    let toRemove = this.scene.children.filter(
+      child => (child.type === 'Mesh' || child.type ===  'Group')
     );
+
+    if (instances) {
+      topInstances = instances.map((i) => { return i.instancePath });
+      for (var i = toRemove.length - 1; i >= 0; i--) {
+        for (var j = 0; j <= topInstances.length - 1; j++) {
+          // if the Mesh is present in the list of instances provided by the canvas then
+          // we can remove this from that list itself and from the toRemove so that we don't recompute
+          // the same mesh/instance twice
+          if (toRemove[i].instancePath.startsWith(topInstances[j])) {
+            // if the color is the same then remove this
```

Comment:
if the color is the same we don't update, differently we do that and the remove the instance. Comment to be updated

Reply:
I'm not a big fan of having a method called clearScene do more than just clearing the scene according to our criteria. (aka I don't like that it also changes colors besides removing stuff)
But I understand the need of having it there.

The first idea that I had was to not remove the instance from instances in this method but rather add some attribute to it so that we could skip the meshFactory work based on that attribute.
But that's just an idea.